group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTestContext.java

```diff
@@ -0,0 +1,1485 @@
+        );
+
+        assertTrue(followerJoinResult.records.isEmpty());
+        assertFalse(followerJoinResult.joinFuture.isDone());
+        assertTrue(group.isInState(PREPARING_REBALANCE));
+        assertEquals(2, group.size());
+        assertEquals(1, group.numPendingJoinMembers());
+
+        return new PendingMemberGroupResult(
+            leaderJoinResponse.memberId(),
+            followerId,
+            pendingMemberJoinResult.joinFuture.get()
+        );
+    }
+
+    public void verifySessionExpiration(ClassicGroup group, int timeoutMs) {
+        Set<String> expectedHeartbeatKeys = group.allMembers().stream()
+                                                 .map(member -> classicGroupHeartbeatKey(group.groupId(), member.memberId())).collect(Collectors.toSet());
+
+        // Member should be removed as session expires.
+        List<MockCoordinatorTimer.ExpiredTimeout<Void, Record>> timeouts = sleep(timeoutMs);
+        List<Record> expectedRecords = Collections.singletonList(newGroupMetadataRecord(
+            group.groupId(),
+            new GroupMetadataValue()
+                .setMembers(Collections.emptyList())
+                .setGeneration(group.generationId())
+                .setLeader(null)
+                .setProtocolType("consumer")
+                .setProtocol(null)
+                .setCurrentStateTimestamp(time.milliseconds()),
+            MetadataVersion.latestTesting()
+        ));
+
+
+        Set<String> heartbeatKeys = timeouts.stream().map(timeout -> timeout.key).collect(Collectors.toSet());
+        assertEquals(expectedHeartbeatKeys, heartbeatKeys);
+
+        // Only the last member leaving the group should result in the empty group metadata record.
+        int timeoutsSize = timeouts.size();
+        assertEquals(expectedRecords, timeouts.get(timeoutsSize - 1).result.records());
+        assertNoOrEmptyResult(timeouts.subList(0, timeoutsSize - 1));
+        assertTrue(group.isInState(EMPTY));
+        assertEquals(0, group.size());
+    }
+
+    public HeartbeatResponseData sendClassicGroupHeartbeat(
+        HeartbeatRequestData request
+    ) {
+        RequestContext context = new RequestContext(
+            new RequestHeader(
+                ApiKeys.HEARTBEAT,
+                ApiKeys.HEARTBEAT.latestVersion(),
+                "client",
+                0
+            ),
+            "1",
+            InetAddress.getLoopbackAddress(),
+            KafkaPrincipal.ANONYMOUS,
+            ListenerName.forSecurityProtocol(SecurityProtocol.PLAINTEXT),
+            SecurityProtocol.PLAINTEXT,
+            ClientInformation.EMPTY,
+            false
+        );
+
+        return groupMetadataManager.classicGroupHeartbeat(
+            context,
+            request
+        );
+    }
+
+    public List<ListGroupsResponseData.ListedGroup> sendListGroups(List<String> statesFilter, List<String> typesFilter) {
+        Set<String> statesFilterSet = statesFilter.stream().collect(Collectors.toSet());
```

Comment:
Can't these be converted to be `new HashSet(statesFilter)` instead of streaming and collecting